tests/runner/test_sequential_runner.py

```diff
@@ -272,6 +272,402 @@ def test_confirms(self, mocker, test_pipeline, is_async):
+        runner = SequentialRunner()
+        catalog = DataCatalog()
+
+        # Test ephemeral dataset (MemoryDataset)
+        mem_dataset = MemoryDataset()
+        catalog._datasets = {"mem_output": mem_dataset}
+        assert not runner._is_persistent_and_missing("mem_output", catalog)
+
+        # Test persistent dataset that exists
+        mock_dataset = mocker.Mock()
+        mock_dataset._EPHEMERAL = False
+        catalog._datasets = {"persist_output": mock_dataset}
+        catalog.exists = mocker.Mock(return_value=True)
+        assert not runner._is_persistent_and_missing("persist_output", catalog)
+
+        # Test persistent dataset that doesn't exist
+        catalog.exists = mocker.Mock(return_value=False)
+        assert runner._is_persistent_and_missing("persist_output", catalog)
+
+        # Test undefined dataset (not in catalog, no factory pattern)
+        catalog._datasets = {}
+        catalog.__contains__ = mocker.Mock(return_value=False)
+        assert not runner._is_persistent_and_missing("undefined_output", catalog)
+
+    def test_only_missing_outputs_factory_pattern_ephemeral(self, mocker, caplog):
+        """Test factory pattern that creates ephemeral datasets"""
+        catalog = DataCatalog()
+
+        # Create a pipeline where output is not defined (will become MemoryDataset)
+        test_pipeline = pipeline([node(identity, "input", "output_test", name="node1")])
+        catalog["input"] = MemoryDataset("test_data")
+
+        runner = SequentialRunner()
+        runner.run(test_pipeline, catalog, only_missing_outputs=True)
+
+        # Should skip everything since undefined outputs become MemoryDatasets (ephemeral)
+        assert "Skipping all 1 nodes (all persistent outputs exist)" in caplog.text
+
+    def test_only_missing_outputs_partial_persistent_outputs(self, mocker, caplog):
+        """Test node with mix of persistent and ephemeral outputs"""
+        catalog = DataCatalog()
+
+        # Node produces both ephemeral and persistent outputs
+        test_pipeline = pipeline(
+            [
+                node(
+                    lambda x: (x, x + "_persistent"),
+                    "input",
+                    ["ephemeral_out", "persistent_out"],
+                    name="mixed_node",
+                ),
+            ]
+        )
+
+        # Setup catalog
+        catalog["input"] = MemoryDataset("data")
+        catalog["ephemeral_out"] = MemoryDataset()
+        catalog["persistent_out"] = LambdaDataset(
+            load=lambda: None, save=lambda x: None
+        )
+
+        # Mock exists for persistent dataset
+        catalog.exists = mocker.Mock(return_value=False)
+
+        runner = SequentialRunner()
+        runner.run(test_pipeline, catalog, only_missing_outputs=True)
+
+        # Should run because persistent output is missing
+        assert "Running all 1 nodes" in caplog.text
+
+    def test_is_output_missing_various_conditions(self, mocker):
```

Comment:
This test is very hard to read. Can't you break each of the separate test cases where there's a comment now into a test?